File: py/singles/polymorphism.py
from math import pi

# base class
class Shape:
    def __init__(self, w, h):
        self.width = w
        self.height = h

    def area(self):
        raise NotImplementedError("Subclasses should implement this!")

# derived classes
class Rectangle(Shape):
    def __init__(self, w, h):
        super().__init__(w, h)

    def area(self):
        return self.width * self.height

class Circle(Shape):
    def __init__(self, w, h):
        super().__init__(w, h)

    def area(self):
        return (self.height / 2) ** 2 * pi

# main functionality
if __name__ == "__main__":
    rect = Rectangle(10, 10)
    circle = Circle(15, 15)
    
    shape = circle 
    print(f" {shape.area()}") # Output will be based on pi value used
    
    shape = rect
    print(f" {shape.area()}")

File: js/singles/polymorphism.js
// base class
class Shape {
  constructor(w, h) {
    this.width = w;
    this.height = h;
  }
  // This should be overridden by derived classes
  area() {
    throw new Error('You have to implement the method area!');
  }
}

// derived classes
class Rectangle extends Shape {
  constructor(w, h) {
    super(w, h);
  }
  area() {
    return this.width * this.height;
  }
}

class Circle extends Shape {
  constructor(w, h) {
    super(w, h);
  }
  area() {
    return (this.height / 2) * (this.height / 2) * Math.PI;
  }
}

// usage
const rect = new Rectangle(10, 10);
const circle = new Circle(15, 15);
let shape = circle; 
console.log(shape.area()); // Output: 176.7145....
shape = rect;
console.log(shape.area()); // Output: 100
